### PR TITLE
Exclude unneeded chunks for IN/ANY statement

### PR DESCRIPTION
**Pull request by niksajakovljevic**
Exclude unneeded chunks for IN/ANY statement.
Optimization only takes care of equals operator. Only explicit values are supported eg.:
`WHERE some_value IN ('val1', 'val2'...) ` or `WHERE some_value = ANY(ARRAY['val1', 'val2'....])`.

IN/ANY with subqueries are not taken care of in this PR.

### CODE REVIEW

**Comment by erimatnor** [comment]
Should make sure you run this through `pgindent`.

File: src/hypertable_restrict_info.c [hunk 1]
```diff
@@ -147,13 +148,27 @@ static DimensionVec *
...
+		foreach(cell, dri->values)
+		{
+			int32		value = lfirst_int(cell);
+			DimensionVec *tmp = dimension_slice_scan_range_limit(dri->base.dimension->fd.id,
+																 BTLessEqualStrategyNumber,
+																 value,
+																 BTGreaterEqualStrategyNumber,
+																 value,
+																 0);
+			for(int i=0; i < tmp->num_slices; i++)
```
Comment (erimatnor) [inline src/hypertable_restrict_info.c]: Mixed code and declaration. Missing whitespace.

File: src/hypertable_restrict_info.c [hunk 2]
```diff
@@ -147,13 +148,27 @@ static DimensionVec *
...
+			{
```
Comment (erimatnor) [inline src/hypertable_restrict_info.c]: Unnecessary braces.

File: src/hypertable_restrict_info.c [hunk 3]
```diff
@@ -293,18 +309,62 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+{
+	ArrayIterator iterator = array_create_iterator(DatumGetArrayTypeP(c->constvalue), 0, NULL);
+	Datum		elem = (Datum) NULL;
+ 	bool		isnull;
+	List	   *values = NIL;
+	
+	while(array_iterate(iterator, &elem, &isnull))
+	{
+		if (!isnull)
+		{
```
Comment (erimatnor) [inline src/hypertable_restrict_info.c]: unnecessary braces.

File: src/hypertable_restrict_info.c [hunk 4]
```diff
@@ -293,18 +309,62 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+
+static List *
+get_scalar_op_expr_values(Const *c)
+{
+	ArrayIterator iterator = array_create_iterator(DatumGetArrayTypeP(c->constvalue), 0, NULL);
+	Datum		elem = (Datum) NULL;
+ 	bool		isnull;
+	List	   *values = NIL;
+	
+	while(array_iterate(iterator, &elem, &isnull))
```
Comment (erimatnor) [inline src/hypertable_restrict_info.c]: missing whitespace

File: src/hypertable_restrict_info.c [hunk 5]
```diff
@@ -147,13 +148,27 @@ static DimensionVec *
...
 		/* slice_end >= value && slice_start <= value */
-		return dimension_slice_scan_range_limit(dri->base.dimension->fd.id,
-												BTLessEqualStrategyNumber,
-												dri->value,
-												BTGreaterEqualStrategyNumber,
-												dri->value,
-												0);
+		ListCell	*cell;
+		DimensionVec *dimVec = dimension_vec_create(DIMENSION_VEC_DEFAULT_SIZE);
+		foreach(cell, dri->values)
```
Comment (erimatnor) [inline src/hypertable_restrict_info.c]: missing whitespace

**Comment by niksajakovljevic** [comment]
@erimatnor Yeah, trying to set it up locally. Requires Perl module....

**Comment by cevian** [comment]
@niksajakovljevic make sure you are using the PG 10.x version of pg_indent

File: src/hypertable_restrict_info.c [hunk 6]
```diff
@@ -105,14 +107,14 @@ dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber
...
+dimension_restrict_info_closed_add(DimensionRestrictInfoClosed *dri, StrategyNumber strategy, Datum datum)
 {
-	int64		value = partitioning_func_apply(dri->base.dimension->partitioning, c->constvalue);
+	int32		value = partitioning_func_apply(dri->base.dimension->partitioning, datum);
 
 	switch (strategy)
 	{
 		case BTEqualStrategyNumber:
-			dri->value = value;
+			dri->values = list_append_unique_int(dri->values, value);
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: I think the list actually needs to be passed down all the way to `dimension_restrict_info_closed_add` since if you have two lists then you want the intersection of the two. Consider `WHERE device_id IN ('dev_1', 'dev_2') and device_id IN ('dev_2', 'dev_3')`

Comment (cevian) [reply src/hypertable_restrict_info.c]: OR would not land either clause in the restrict info I believe. You should double check. I think restrict info are only clause that can be anded together at the top level

Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Actually that is one of the things we might want to address. If you have some boolean expression like above we will not add any restriction info so we would scan all the chunks. Shouldn't we try to expand boolean expressions?

File: src/hypertable_restrict_info.c [hunk 7]
```diff
@@ -293,18 +348,56 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+	bool		added;
+
 	Expr	   *e = ri->clause;
 
-	if (!IsA(e, OpExpr))
+	/* Same as constraint_exclusion */
+	if (contain_mutable_functions((Node *) e))
 		return;
+	added = false;
+	if (IsA(e, OpExpr))
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: should this be a switch on the type instead of an else/if?

File: src/hypertable_restrict_info.c [hunk 8]
```diff
@@ -147,13 +189,28 @@ static DimensionVec *
...
-												BTGreaterEqualStrategyNumber,
-												dri->value,
-												0);
+		ListCell   *cell;
+		DimensionVec *dimVec = dimension_vec_create(DIMENSION_VEC_DEFAULT_SIZE);
+
+		foreach(cell, dri->partitions)
+		{
+			int			i;
+			int32		parition = lfirst_int(cell);
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: spelling of parition->partition

File: src/hypertable_restrict_info.c [hunk 9]
```diff
@@ -71,68 +73,108 @@ dimension_restrict_info_create(Dimension *d)
 }
 
 static bool
-dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, Const *c)
+dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, List *values, Oid type)
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: This does not seem correct to me. It again handles the OR case the same as the AND case and this isn't right. `WHERE time < '2001-01-01' AND time < '2000-01-01'` seems like a different case than `WHERE time < ANY ('2001-01-01', '2000-01-01')`. But this code will treat it equivalently?

Comment (cevian) [reply src/hypertable_restrict_info.c]: Related question how does this code handle differences between ANY and ALL

File: src/hypertable_restrict_info.c [hunk 10]
```diff
@@ -71,68 +73,108 @@ dimension_restrict_info_create(Dimension *d)
...
+		switch (strategy)
+		{
+			case BTLessEqualStrategyNumber:
+			case BTLessStrategyNumber:
+				if (dri->upper_strategy == InvalidStrategy || value < dri->upper_bound)
+				{
+					dri->upper_strategy = strategy;
+					dri->upper_bound = value;
+				}
+				restriction_added |= true;
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Bitwise OR looks more elegant (at least to me :) but I agree it can bring confusion since we only operate on boolean type so I'll change it. 
The existing code was returning `true` outside a curly braces so I kept the same behaviour. Maybe @cevian can shed some light there?

File: src/hypertable_restrict_info.c (hunk 9 above)
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Actually this PR is not addressing open dimensions like time. I think we went through it already and decided not to support IN/ANY for time dimension. IN/ANY/ALL are all handled in the same way.

File: src/hypertable_restrict_info.c [hunk 11]
```diff
@@ -71,68 +73,108 @@ dimension_restrict_info_create(Dimension *d)
...
 static bool
-dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, Const *c)
+dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, List *values, Oid type)
 {
-	int64		value = time_value_to_internal(c->constvalue, c->consttype, false);
+	ListCell   *item;
+	bool		restriction_added;
 
-	switch (strategy)
+	restriction_added = false;
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: The reason I had it separate is b/c I thought `pg_indent` complained about it / was wrong about it.

File: src/hypertable_restrict_info.c [hunk 12]
```diff
@@ -71,68 +73,119 @@ dimension_restrict_info_create(Dimension *d)
...
+
+			if (!use_or && list_length(dri->partitions) > 1)
+				/* since this is AND the intersection is empty set */
+			{
+				dri->partitions = NIL;
+				return restriction_added;
+			}
+		}
+	}
+	else
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: how come use_or is not used in the else?

Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Good catch!

File: src/hypertable_restrict_info.c [hunk 13]
```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
...
 
-	switch (strategy)
+	if (strategy != BTEqualStrategyNumber)
 	{
-		case BTEqualStrategyNumber:
-			dri->value = value;
+		return restriction_added;
+	}
+
+	if (dri->partitions == NIL)
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: I think this is the wrong check dri->partitions can be NIL in 2 cases:
1) dri->partitions was never set because the dri was just created.
2) all partitions were excluded (the list is empty)
I think it's incorrect to conflate the two.

 I think the test you want here is dri->strategy == InvalidStrategy

File: src/hypertable_restrict_info.c [hunk 14]
```diff
@@ -220,48 +297,46 @@ hypertable_restrict_info_get(HypertableRestrictInfo *hri, AttrNumber attno)
 	return NULL;
 }
 
+typedef List *(*get_const_expr_values) (Const *c);
+
 static bool
-hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *root, OpExpr *clause)
+hypertable_restrict_info_add_expr(HypertableRestrictInfo *hri, PlannerInfo *root, List *expr_args, Oid op_oid, get_const_expr_values const_values, bool use_or)
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: I think passing const_values here makes this harder to read than necessary. Can't we change this function to return a const * (or NULL) and have the calling function (hypertable_restrict_info_add_restrict_info) call either get_scalar_op_expr_values or get_op_expr_values as appropriate since it knows which one to call?

File: src/hypertable_restrict_info.c [hunk 15]
```diff
@@ -293,18 +368,66 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+
+				added = hypertable_restrict_info_add_expr(hri, root, op_expr->args, op_expr->opno, get_op_expr_values, false);
+				break;
+			}
+
+		case T_ScalarArrayOpExpr:
+			{
+				ScalarArrayOpExpr *scalar_expr = (ScalarArrayOpExpr *) e;
+
+				added = hypertable_restrict_info_add_expr(hri, root, scalar_expr->args, scalar_expr->opno, get_scalar_op_expr_values, scalar_expr->useOr);
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: another possible way to simplify this code: isn't it true that if scalar_expr->useOr == false, we can just iterate over the list here and call dimension_restrict_info_add on each individual list element. This is because conceptually ` a op 'id1' and a op 'id2'` is the same as `a op ALL('id1' and 'id2')` but right now these take vastly different code paths. Ideally, I think `dimension_restrict_info_add` should only take lists of items that correspond to "OR" lists. For AND lists we can just call dimension_restrict_info_add in a loop with the elements in the list.

File: src/hypertable_restrict_info.c [hunk 16]
```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
...
-			return true;
-		case BTGreaterEqualStrategyNumber:
-		case BTGreaterStrategyNumber:
-			if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
-			{
-				dri->lower_strategy = strategy;
+		Datum		datum = PointerGetDatum(lfirst(item));
+		int64		value = time_value_to_internal(datum, type, true);
+
+		if (value == -1)
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Basically `time_value_to_internal` returns `-1` when in can't convert Datum type to value. IN/ANY/ALL has specific Oid type that we can't convert.

File: src/hypertable_restrict_info.c (hunk 9 above)
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: We filter out ANY/ALL/IN here (if we get `-1` when trying to convert Datum to int64) https://github.com/timescale/timescaledb/pull/784/commits/f14c8414c21678aa1512d1e6682f8068b5fa66b9#diff-273cc5022b4d73507f7284cbee02026bR98

File: src/hypertable_restrict_info.c [hunk 17]
```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
...
+	else
+	{
+		List	   *tmp = NIL;
+
+		foreach(item, values)
+		{
+			Datum		datum = PointerGetDatum(lfirst(item));
+			int32		partition = partitioning_func_apply(dri->base.dimension->partitioning, datum);
+
+			tmp = lappend_int(tmp, partition);
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: Yes, since we iterate over unique `dri->partitions` and check existence in `tmp`. I've skipped unique check here because current PostgreSQL impl. is quadratic time complexity

File: src/hypertable_restrict_info.c (hunk 15 above)
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: I am not sure if for certain cases we can directly call `dimension_restrict_info_add ` here b/c there are couple of checks we need to run before ( eg. various checks in `hypertable_restrict_info_add_expr`)
I wouldn't say that `a op 'id1' AND a op 'id2'` and `a op ALL('id1','id2')` take vastly different code paths: the only difference between these two is how you extract values from `Const`.

File: src/hypertable_restrict_info.c (hunk 14 above)
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: I see two problems with that approach: 1. there are various checks and transformations that happen in `hypertable_restrict_info_add_expr` before we can actually try to extract the values 2. we will upfront pay the price for extracting the values even if we fail on some checks after

File: src/hypertable_restrict_info.c [hunk 18]
```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
...
+		}
+		dri->partitions = list_intersection_int(dri->partitions, tmp);
+		/* no intersection is also a restriction  */
+		restriction_added = true;
+		if (remove_all_partitions(use_or, dri))
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: I refactored this part of a code. Let me know if it's more intuitive now

File: src/hypertable_restrict_info.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable_restrict_info.c]: But this is a value inside the IN/ANY/ALL, no. Why would it have the OID of the in/any/all

[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: > But this is a value inside the IN/ANY/ALL, no. Why would it have the OID of the in/any/all

We pass OID we've got from `Const`. In case of IN/ANY/ALL that's array type. Since we agreed not to optimize time with IN/ANY, I took this shortcut. Alternatively I was thinking to check if OID is array type and return false immediately ( it looked a bit ugly to copy/paste all array OIDs somewhere and check against, especially since we already have a method that check if it's valid time type). I understand that his look a bit weird but I am not strongly on this. Thoughts?

> why don't we just return false from the function at the beginning if(dimValues->useOr && list_length(dimValues->values) > 1)

We can, but it's been taken care of after `if (value == -1)` check. If `list_length(dimValues->values) > 1` means that OID type is array and we will return after trying to convert to value.

[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable_restrict_info.c]: Sorry what I meant was `if(dimValues->useOr && list_length(dimValues->values) > 1)` before the foreach and then a value == -1 becomes an error. Also the type can go into the DimensionValues and I would make the type be the base value type, not the corresponding array type. I think this is more correct and reads better. It also handle the ANY(one_item) case which is a plus.

[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable_restrict_info.c]: Also my proposal would handle the ALL case naturally which is also another plus.

[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply src/hypertable_restrict_info.c]: I like the idea of wrapping type with dim values. To process ANY/ALL we would first need to convert array type into non-array type: that's easy to do but it will add additional `switch` with few hardcoded Oids (not so nice since it's a weird corner case). However  I am not strongly opinionated about it.  Thoughts?

File: test/expected/plan_expand_hypertable_optimized.out [hunk 19]
```diff
@@ -421,10 +433,242 @@ SELECT * FROM cte ORDER BY value;
...
+   ->  Append
+         ->  Seq Scan on hyper_w_space
+               Filter: (("time" < 10) AND (device_id = ANY ('{dev5,dev6}'::text[])))
+         ->  Seq Scan on _hyper_2_106_chunk
+               Filter: (("time" < 10) AND (device_id = ANY ('{dev5,dev6}'::text[])))
+         ->  Seq Scan on _hyper_2_105_chunk
+               Filter: (("time" < 10) AND (device_id = ANY ('{dev5,dev6}'::text[])))
+(9 rows)
+
+--ANY with intersection
```
Comment (cevian) [inline test/expected/plan_expand_hypertable_optimized.out]: let's add a test with ANY with intersection, where that intersection should be empty.

File: test/expected/plan_expand_hypertable_optimized.out [hunk 20]
```diff
@@ -421,10 +433,242 @@ SELECT * FROM cte ORDER BY value;
...
+         ->  Seq Scan on _hyper_2_104_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+         ->  Seq Scan on _hyper_2_105_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+         ->  Seq Scan on _hyper_2_103_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+(13 rows)
+
+--Time dimension doesn't filter chunks when used with IN/ANY/ALL
+:PREFIX SELECT * FROM hyper_w_space WHERE time < ANY(ARRAY[1,2]) ORDER BY value;
```
Comment (cevian) [inline test/expected/plan_expand_hypertable_optimized.out]: time should now work with ANY(one_element) and ALL(one or many). Let's add tests for those cases.

File: test/expected/plan_expand_hypertable_optimized.out [hunk 21]
```diff
@@ -421,10 +433,242 @@ SELECT * FROM cte ORDER BY value;
...
+               Filter: ("time" < ANY ('{1,2}'::integer[]))
+         ->  Seq Scan on _hyper_2_113_chunk
+               Filter: ("time" < ANY ('{1,2}'::integer[]))
+         ->  Seq Scan on _hyper_2_114_chunk
+               Filter: ("time" < ANY ('{1,2}'::integer[]))
+         ->  Seq Scan on _hyper_2_115_chunk
+               Filter: ("time" < ANY ('{1,2}'::integer[]))
+(31 rows)
+
+--AND intersection with IN and EQUALS
```
Comment (cevian) [inline test/expected/plan_expand_hypertable_optimized.out]: again let's add a test for empty intersections.

File: src/hypertable_restrict_info.c [hunk 22]
```diff
@@ -299,18 +388,84 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
 							   &lefttype,
 							   &righttype);
 
-	return dimension_restrict_info_add(dri, strategy, c);
+	dimValues = func_get_dim_values(c, use_or);
+	return dimension_restrict_info_add(dri, strategy, dimValues);
+}
+
+static DimensionValues *
+create_dimension_values(List *values, Oid const_type, bool use_or)
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: should be `dimension_values_create`

File: src/hypertable_restrict_info.c [hunk 23]
```diff
@@ -299,18 +388,84 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+{
+	Oid			base_el_type;
+	DimensionValues *dimValues;
+
+	dimValues = palloc(sizeof(DimensionValues));
+	dimValues->values = values;
+	dimValues->use_or = use_or;
+	/* if an array type, lets get the base element type */
+	base_el_type = get_element_type(const_type);
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: I think this logic should be in `dimension_values_create_from_array` and it should be an error if it fails. This kind of code where you assume a reason for a failure is always more brittle. In  `dimension_values_create_from_array` we know it should be an array.

File: src/hypertable_restrict_info.c [hunk 24]
```diff
@@ -299,18 +388,84 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
...
+
+static DimensionValues *
+dimension_values_create_from_array(Const *c, bool user_or)
+{
+	ArrayIterator iterator = array_create_iterator(DatumGetArrayTypeP(c->constvalue), 0, NULL);
+	Datum		elem = (Datum) NULL;
+	bool		isnull;
+	List	   *values = NIL;
+
+	while (array_iterate(iterator, &elem, &isnull))
```
Comment (cevian) [inline src/hypertable_restrict_info.c]: please put braces around the while.

File: test/expected/plan_expand_hypertable_optimized.out [hunk 25]
```diff
@@ -421,10 +433,302 @@ SELECT * FROM cte ORDER BY value;
...
+               Filter: (("time" < 10) AND ("time" < 100))
+         ->  Seq Scan on _hyper_2_104_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+         ->  Seq Scan on _hyper_2_105_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+         ->  Seq Scan on _hyper_2_103_chunk
+               Filter: (("time" < 10) AND ("time" < 100))
+(13 rows)
+
+--Time dimension doesn't filter chunks when using IN/ANY/ALL with multiple arguments
```
Comment (cevian) [inline test/expected/plan_expand_hypertable_optimized.out]: update comment to take out ALL

File: test/expected/plan_expand_hypertable_optimized.out [hunk 26]
```diff
@@ -421,10 +433,302 @@ SELECT * FROM cte ORDER BY value;
...
+         ->  Seq Scan on _hyper_2_104_chunk
+               Filter: ("time" < ANY ('{1}'::integer[]))
+         ->  Seq Scan on _hyper_2_105_chunk
+               Filter: ("time" < ANY ('{1}'::integer[]))
+         ->  Seq Scan on _hyper_2_103_chunk
+               Filter: ("time" < ANY ('{1}'::integer[]))
+(13 rows)
+
+--Time dimension chunk filtering works for ALL with single argument
+:PREFIX SELECT * FROM hyper_w_space WHERE time < ALL(ARRAY[1]) ORDER BY value;
```
Comment (cevian) [inline test/expected/plan_expand_hypertable_optimized.out]: all should also work with multiple elements please add test.

**Comment by niksajakovljevic** [comment]
@cevian Squashed commits and improved commit message.
